src/Network/WebSockets/Routing.hs

```diff
@@ -0,0 +1,207 @@
+subprotocol :: ByteString -> WebSocketsRoute a -> WebSocketsRoute a
+subprotocol proto w = do
+    p <- askPending'
+    let reqProto = getRequestSubprotocols . pendingRequest $ p
+    if proto `elem` reqProto then
+        setProto (Just proto) w
+      else
+        mzero
+
+-- | Guard which only succeeds if no subprotocols have been requested.
+noSubprotocols :: WebSocketsRoute ()
+noSubprotocols = do
+    p <- askPending'
+    guard $ null . getRequestSubprotocols . pendingRequest $ p
+
+--------------------------------------------------------------------------------
+-- Route by path info
+
+-- | Pop a path element and run the supplied handler if it matches the given
+-- string.
+--
+-- > route :: WebSocketsRoute ()
+-- > route = dir "foo" $ dir "bar" $ subRoute
+--
+-- The path element can not contain \'/\'. See also 'dirs'.
+dir :: String -> WebSocketsRoute a -> WebSocketsRoute a
+dir p r = do
+    pending <- askPending'
+    case paths pending of
+        (p':ps) | p == dropTrailingPathSeparator p' -> do
+            let ps' | hasTrailingPathSeparator p' = "/" : ps
+                    | otherwise                   =       ps
+            withPending (setPath ps') r
+        _ -> mzero
+
+-- | Guard against a 'FilePath'. Unlike 'dir' the 'FilePath' may contain \'/\'.
+-- If the guard succeeds, the matched elements will be popped from the directory stack.
+--
+-- > dirs "foo/bar" $ ...
+--
+-- See also: 'dir'.
+dirs :: FilePath -> WebSocketsRoute a -> WebSocketsRoute a
+dirs fp w = do
+    let parts = splitDirectories $ makeRelative "/" fp
+    foldr dir w parts
+
+-- | Guard which only succeeds if there are no remaining path segments. Often
+-- used if you want to explicitly assign a route for \'/\'.
+nullDir :: WebSocketsRoute ()
+nullDir = do
+    pending <- askPending'
+    guard $ null $ paths pending
+
+-- | The opposite of 'nullDir': Succeeds if there are path segments left, fails
+-- otherwise.
+notNullDir :: WebSocketsRoute ()
+notNullDir = do
+    pending <- askPending'
+    guard $ not . null $ paths pending
+
+-- | Guard which checks that the request URI ends in \'/\'. Useful for
+-- distinguishing between @foo@ and @foo/@.
+trailingSlash :: WebSocketsRoute ()
+trailingSlash = do
+    pending <- askPending'
+    case B8.unsnoc . requestPath . pendingRequest $ pending of
+        Just (_, '/') -> return ()
+        _             -> mzero
+
+-- | The opposite of 'trailingSlash'.
+noTrailingSlash :: WebSocketsRoute ()
+noTrailingSlash = do
+    pending <- askPending'
+    case B8.unsnoc . requestPath . pendingRequest $ pending of
+        Just (_, '/') -> mzero
+        _             -> return ()
+
+-- | Pop any path element and run the websocket route. Succeeds if a path
+-- component was popped. Fails if the remaining path was empty.
+path :: (String -> WebSocketsRoute a) -> WebSocketsRoute a
+path r = do
+    pending <- askPending'
+    case paths pending of
+        (p:ps) | p /= "." -> -- makeRelative turns "/" into ["."]
```

Comment:
In this whole file, I think it would be beneficial to wrap the path (de-)composition logic, to better separate concerns, and make the code more readable. For example here, instead of `p /= "."` with a comment as to why we don't compare with "/", I would prefer seeing `not empty p` or something like that.